### PR TITLE
Move cache cleanup operation on state modification operation and make…

### PR DESCRIPTION
**Pull request by kandy**
### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

**Comment by magento-cicd2** [comment]
[CLA assistant check](https://cla.dev.magento.com/magento/magento2?pullRequest=22752) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dev.magento.com/magento/magento2?pullRequest=22752) before we can accept your contribution.<br/>

File: lib/internal/Magento/Framework/Filesystem/Driver/StatefulFile.php [hunk 1]
```diff
@@ -0,0 +1,1005 @@
...
+use Magento\Framework\Filesystem\Glob;
+
+/**
+ * Filesystem driver that uses the local filesystem.
+ * Assumed that stat cache is cleanup before test filesystem
+ *
+ * @package Magento\Framework\Filesystem\Driver
+ * @SuppressWarnings(PHPMD.ExcessiveClassComplexity)
+ */
+class StatefulFile implements DriverInterface
```
Comment (orlangur) [inline lib/internal/Magento/Framework/Filesystem/Driver/StatefulFile.php:22]: Is it just and old class copied without modification?

File: lib/internal/Magento/Framework/Filesystem/Driver/StatefulFile.php [hunk 2]
```diff
@@ -0,0 +1,1005 @@
+<?php
+/**
+ * Origin filesystem driver
```
Comment (orlangur) [inline lib/internal/Magento/Framework/Filesystem/Driver/StatefulFile.php:3]: Move comment to a better place and add strict types, please.

**Review by orlangur — CHANGES_REQUESTED** [review]
@kandy please provide a kind of steps for smoke testing.

File: lib/internal/Magento/Framework/Filesystem/Driver/StatefulFile.php (hunk 1 above)
Comment (kandy) [reply lib/internal/Magento/Framework/Filesystem/Driver/StatefulFile.php:22]: Mostly yes, except the updates to the clearstatcache. Now it cleanup only requested file (see second params)

Comment (orlangur) [reply lib/internal/Magento/Framework/Filesystem/Driver/StatefulFile.php:22]: @kandy ok, then I'll skip parts like new `protected` properties/methods and other code smell.

**Comment by sidolov** [comment]
Hi @kandy , will you continue progress on the PR?

**Comment by kandy** [comment]
@sidolov, Yep, too busy now

**Comment by sidolov** [comment]
Hi @kandy , I am closing this PR now due to inactivity.
Please reopen and update pull request when you will have time to continue.
Thank you for the collaboration!

**Pull request closed without merging**